### PR TITLE
feat: vbase optimization

### LINKED ISSUE #171: feat: Post-filter with VBASE

### PR DESCRIPTION
**Pull request by silver-ymz**
close #171

### CODE REVIEW

**Comment by gaocegege** [comment]
Thanks for your contribution! :tada: :+1:

File: src/algorithms/hnsw.rs [hunk 1]
```diff
@@ -509,6 +531,55 @@ pub fn local_search(
...
+            if !visited.check(v) {
+                continue;
+            }
+            visited.mark(v);
+            let v_dis = mmap.quantization.distance(mmap.d, vector, v);
+            if !results.check(v_dis) {
+                continue;
+            }
+            candidates.push(Reverse((v_dis, v)));
+            results.push(HeapElement {
```
Comment (usamoi) [inline src/algorithms/hnsw.rs:567]: Will it lost data to be returned?

Comment (silver-ymz) [reply src/algorithms/hnsw.rs:567]: Yes.

Comment (usamoi) [reply src/algorithms/hnsw.rs:567]: The first stage points should be saved and sorted. In the second stage, we will return a better choice in first stage points and second stage heap.

Comment (usamoi) [reply src/algorithms/hnsw.rs:567]: Not all, at most `range` points here.

Comment (silver-ymz) [reply src/algorithms/hnsw.rs:567]: The line 578 saves the the first stage heap to the iterator. The current implementation has already chosen the first stage point.

Comment (usamoi) [reply src/algorithms/hnsw.rs:567]: First stage heap contains at most `range` points. All points in first stage, including points that are not in first stage heap, should be returned.

Comment (silver-ymz) [reply src/algorithms/hnsw.rs:567]: What do you mean by the points in the first stage but not in the first stage heap?

Comment (usamoi) [reply src/algorithms/hnsw.rs:567]: All visited points. `results` will contain at most `range` points, and points are deleted by `heap::push` are lost.

File: src/algorithms/hnsw.rs [hunk 2]
```diff
@@ -509,6 +531,61 @@ pub fn local_search(
...
+    assert!(range > 0);
+    let mut visited_guard = mmap.visited.fetch();
+    let mut visited = visited_guard.fetch();
+    let mut candidates = BinaryHeap::<Reverse<(Scalar, u32)>>::new();
+    let mut results = Heap::new(range);
+    let mut heap = BinaryHeap::<Reverse<HeapElement>>::new();
+    visited.mark(s);
+    let s_dis = mmap.quantization.distance(mmap.d, vector, s);
+    candidates.push(Reverse((s_dis, s)));
+    results.push(HeapElement {
```
Comment (usamoi) [inline src/algorithms/hnsw.rs:549]: Push to `heap`.

A vec to be sorted is better here.

Comment (usamoi) [reply src/algorithms/hnsw.rs:549]: Push `s` to heap.

`heap` need to be sorted.

Compare head of `heap` and `result` to return the next result .

Comment (silver-ymz) [reply src/algorithms/hnsw.rs:549]: Fixed in new commit. Could you review again?

**Pull request merged**